src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java

```diff
@@ -250,52 +255,45 @@ String computeCrc32c(String input) {
         return outputs;
     }
 
-    static boolean shouldRetry(Throwable throwable) {
+    // For batch operations, we want to also include the AbortedException as retryable
+    static boolean shouldRetry(Throwable throwable, boolean isBatch) {
         if (throwable == null) {
             return false; // Did not find a retry-able exception
         }
         if (throwable instanceof DeadlineExceededException ||
             throwable instanceof UnavailableException ||
-            throwable instanceof AbortedException ||
             throwable instanceof InternalException ||
-            throwable instanceof StatusRuntimeException) {
-
+            throwable instanceof StatusRuntimeException ||
+            (isBatch && throwable instanceof AbortedException)) {
             return true;
         }
-        return shouldRetry(throwable.getCause());
+        return shouldRetry(throwable.getCause(), isBatch);
     }
 
-    // TODO - convert this to a generic, remove duplicate code from FireStoreDirectoryDao
-    interface FirestoreFunction {
-        void apply(Transaction xn) throws InterruptedException;
-    }
-
-    public void runTransactionWithRetry(Firestore firestore,
-                                        FirestoreFunction firestoreFunction,
-                                        String transactionOp,
-                                        String warnMessage) throws InterruptedException {
+    public <T> T runTransactionWithRetry(Firestore firestore,
+                                         Transaction.Function<T> firestoreFunction,
+                                         String transactionOp,
+                                         String warnMessage) throws InterruptedException {
         int retry = 0;
         while (true) {
             try {
-                ApiFuture<Void> transaction =
-                    firestore.runTransaction(
-                        xn -> {
-                            firestoreFunction.apply(xn);
-                            return null;
-                        });
+                ApiFuture<T> transaction = firestore.runTransaction(firestoreFunction);
 
-                transactionGet(transactionOp, transaction);
-                return;
+                return transactionGet(transactionOp, transaction);
             } catch (Exception ex) {
                 final long retryWait = SLEEP_BASE_MILLISECONDS * Double.valueOf(Math.pow(2.5, retry)).longValue();
-                if (FireStoreUtils.shouldRetry(ex) && retry < firestoreRetries) {
+                if (FireStoreUtils.shouldRetry(ex, false) && retry < firestoreRetries) {
```

Comment:
A minor thing, I would do the `<` operation first since `&&` short circuits and I'd bet `<` is faster than `FireStoreUtils.shouldRetry()`. 
```suggestion
                if (retry < firestoreRetries && FireStoreUtils.shouldRetry(ex, false)) {
```